exchange/bitswap/session.go

```diff
@@ -0,0 +1,233 @@
+	liveCnt   int
+
+	tick          *time.Timer
+	baseTickDelay time.Duration
+
+	latTotal time.Duration
+	fetchcnt int
+
+	notif notifications.PubSub
+
+	uuid logging.Loggable
+
+	id uint64
+}
+
+// NewSession creates a new bitswap session whose lifetime is bounded by the
+// given context
+func (bs *Bitswap) NewSession(ctx context.Context) *Session {
+	s := &Session{
+		activePeers:   make(map[peer.ID]struct{}),
+		liveWants:     make(map[string]time.Time),
+		newReqs:       make(chan []*cid.Cid),
+		cancelKeys:    make(chan []*cid.Cid),
+		ctx:           ctx,
+		bs:            bs,
+		incoming:      make(chan blkRecv),
+		notif:         notifications.New(),
+		uuid:          loggables.Uuid("GetBlockRequest"),
+		baseTickDelay: time.Millisecond * 500,
+		id:            bs.getNextSessionID(),
+	}
+
+	cache, _ := lru.New(2048)
+	s.interest = cache
+
+	bs.sessLk.Lock()
+	bs.sessions = append(bs.sessions, s)
+	bs.sessLk.Unlock()
+
+	go s.run(ctx)
+
+	return s
+}
+
+type blkRecv struct {
+	from peer.ID
+	blk  blocks.Block
+}
+
+func (s *Session) receiveBlockFrom(from peer.ID, blk blocks.Block) {
+	s.incoming <- blkRecv{from: from, blk: blk}
+}
+
+func (s *Session) interestedIn(c *cid.Cid) bool {
+	return s.interest.Contains(c.KeyString())
+}
+
+const provSearchDelay = time.Second * 10
+
+func (s *Session) addActivePeer(p peer.ID) {
+	if _, ok := s.activePeers[p]; !ok {
+		s.activePeers[p] = struct{}{}
+		s.activePeersArr = append(s.activePeersArr, p)
+	}
+}
+
+func (s *Session) resetTick() {
+	if s.latTotal == 0 {
+		s.tick.Reset(provSearchDelay)
+	} else {
+		avLat := s.latTotal / time.Duration(s.fetchcnt)
+		s.tick.Reset(s.baseTickDelay + (3 * avLat))
+	}
+}
+
+func (s *Session) run(ctx context.Context) {
+	s.tick = time.NewTimer(provSearchDelay)
+	newpeers := make(chan peer.ID, 16)
+	for {
+		select {
+		case blk := <-s.incoming:
+			s.tick.Stop()
+
+			s.addActivePeer(blk.from)
+
+			s.receiveBlock(ctx, blk.blk)
+
+			s.resetTick()
+		case keys := <-s.newReqs:
+			for _, k := range keys {
+				s.interest.Add(k.KeyString(), nil)
+			}
+			if s.liveCnt < activeWantsLimit {
+				toadd := activeWantsLimit - s.liveCnt
+				if toadd > len(keys) {
+					toadd = len(keys)
+				}
+				s.liveCnt += toadd
+
+				now := keys[:toadd]
+				keys = keys[toadd:]
+
+				s.wantBlocks(ctx, now)
+			}
+			s.tofetch = append(s.tofetch, keys...)
+		case keys := <-s.cancelKeys:
+			s.cancel(keys)
+
+		case <-s.tick.C:
+			var live []*cid.Cid
+			for c := range s.liveWants {
+				cs, _ := cid.Cast([]byte(c))
+				live = append(live, cs)
+				s.liveWants[c] = time.Now()
+			}
+
+			// Broadcast these keys to everyone we're connected to
+			s.bs.wm.WantBlocks(ctx, live, nil, s.id)
+
+			if len(live) > 0 {
+				go func(k *cid.Cid) {
+					for p := range s.bs.network.FindProvidersAsync(ctx, k, 10) {
+						newpeers <- p
+					}
+				}(live[0])
+			}
+			s.resetTick()
+		case p := <-newpeers:
+			s.addActivePeer(p)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
```

Comment:
I agree, I just think it may be worth waiting a few ms *before* spamming the entire session if we already *know* of some potential providers (i.e., we *don't* have to ask the DHT). 

I still think the current session-centric architecture may make this a bit difficult but it'll probably be easier to just fix it later than to try to hash out all the issues up-front. For all I know, I'm totally wrong and arguing about a non-existent issue.